mlflow/mlflow
# Add universal mlflow.autolog() function!

## What changes are proposed in this pull request?

- Add universal autolog function that enables autologging for the libraries that have been or will be imported.
- Modify spark autologging to not require an active SparkContext when it is initialized -- instead, patch the function to wait for one to be created. This way, an exception will not be thrown when invoking mlflow.autolog() (since pyspark is often imported in our own mlflow code, hence activating autolog for it)

## How is this patch tested?

- Manual testing
- Unit tests
    - verify the autolog function of each library is invoked after calling universal autolog and importing the library
    - verify spark autologging does not throw anymore when a SparkContext is missing
    - verify spark autologging attaches correctly to a SparkContext created AFTER spark autologging was initialized

## Release Notes

### Is this a user-facing change?


Add a universal autolog function mlflow.autolog(), allowing you to make one autolog call and automatically activate all MLflow's supported autologging integrations.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting

Language 

Integrations

### How should the PR be classified in the release notes? Choose one:
